File: code/negative_binomial/parameter.py

# 負の二項分布 ------------------------------------------------------------------

# パラメータの可視化


# %%

# ライブラリの読込 ---------------------------------------------------------------

# ライブラリを読込
import numpy as np
from scipy.stats import nbinom, norm
import matplotlib.pyplot as plt
from matplotlib.animation import FuncAnimation


# %%

# パラメータの影響 ---------------------------------------------------------------

### パラメータの設定 -----

# フレーム数を指定
frame_num = 101

# フレームごとのパラメータを指定
r_vals   = np.linspace(start=0, stop=100, num=frame_num, dtype='int')
phi_vals = np.linspace(start=0.5, stop=0.5, num=frame_num)


# %%

### 変数の設定 -----

# x軸の範囲を設定
k = 1
u = 5.0
x_min = 0
tmp_vals = r_vals * (1.0-phi_vals) / phi_vals # 基準値を指定
x_max = np.max(tmp_vals[np.isfinite(tmp_vals)])
x_max *= k # 定数倍
x_max = np.ceil(x_max /u)*u # u単位で切り上げ
x_max = x_max.astype(np.int64) # 整数型に変換
print('x size:', x_min, x_max)

# x軸の余白を指定:(「モーメントとの関係」用)
x_margin = 0
x_min -= x_margin
x_max += x_margin

# x軸の値を作成
x_vec = np.arange(start=x_min, stop=x_max+1, step=1)


# %%

### 分布の計算 -----

# 負の二項分布の確率を計算
prob_lt = [
    nbinom.pmf(k=x_vec, n=r_vals[i], p=phi_vals[i]) for i in range(frame_num)
]


# %%

### 分布の作図 -----

# 確率軸の範囲を設定
u = 0.05
tmp_arr  = np.array(prob_lt)
prob_max = np.max(tmp_arr[np.isfinite(tmp_arr)])
prob_max = np.ceil(prob_max /u)*u # u単位で切り上げ
print('p(x) size:', prob_max)


# %%

#### パラメータと形状の関係 -----

# 図を初期化
fig, ax = plt.subplots(figsize=(9, 6), dpi=100, facecolor='white')
fig.suptitle('Negative Binomial distribution', fontsize=20)

# 初期化処理を定義
def init():
    pass

# 作図処理を定義
def update(i):

    # 前フレームのグラフを初期化
    ax.cla()

    # 値を取得
    r   = r_vals[i]   # 成功回数パラメータ
    phi = phi_vals[i] # 成功確率パラメータ
    prob_vec = prob_lt[i] # 確率

    # ラベル用の文字列を作成
    param_lbl = f'$r = {r}, \\phi = {phi:.2f}$'
    
    # 負の二項分布を描画
    ax.bar(
        x=x_vec, height=prob_vec, 
        color='#00A968'
    ) # 確率
    #ax.set_xticks(ticks=x_vec) # x軸目盛
    ax.set_xlabel('$x$')
    ax.set_ylabel('probability')
    ax.set_title(param_lbl, loc='left')
    ax.grid()
    ax.set_ylim(ymin=0.0, ymax=prob_max) # 描画範囲を固定

# 動画を作成
anim = FuncAnimation(
    fig=fig, func=update, init_func=init, 
    frames=frame_num, interval=100
)

# 動画を書出
anim.save(
    filename='../../figure/negative_binomial/parameter/parameter.mp4', 
    progress_callback=lambda i, n: print(f'\rframe: {i+1} / {n}', end='', flush=True)
)


# %%

#### パラメータと統計量の関係 -----

# ラベルの表示用の余白を設定
y_margin = 0.05
y_min = -prob_max * y_margin
y_max = prob_max * (1.0+y_margin)

# 図を初期化
fig, ax = plt.subplots(figsize=(9, 6), dpi=100, facecolor='white')
fig.suptitle('Negative Binomial distribution', fontsize=20)

# 初期化処理を定義
def init():
    pass

# 作図処理を定義
def update(i):

    # 前フレームのグラフを初期化
    ax.cla()

    # 値を取得
    r   = r_vals[i]   # 成功回数パラメータ
    phi = phi_vals[i] # 成功確率パラメータ
    prob_vec = prob_lt[i] # 確率

    # 統計量を計算
    if phi > 0.0:
        mean_x = r * (1.0-phi) / phi               # 期待値
        sd_x   = np.sqrt(r * (1.0-phi) / phi**2)   # 標準偏差
        mode_x = np.floor((r-1) * (1.0-phi) / phi) # 最頻値
        mode_x = 1.0 if r <= 1 else mode_x
    else: # 0除算を回避
        mean_x = np.inf
        sd_x   = np.inf
        mode_x = np.inf

    # ラベル用の文字列を作成
    param_lbl = f'$r = {r}, \\phi = {phi:.2f}$'
    
    # 負の二項分布を描画
    ax.bar(
        x=x_vec, height=prob_vec, 
        color='#00A968', alpha=0.5, 
        zorder=10
    ) # 確率
    ax.axvline(
        x=mean_x, 
        color='black', linewidth=1.0, linestyle='--', 
        label=f'$E[x] = \\frac{{r (1-\phi)}}{{\phi}} = {mean_x:.2f}$', 
        zorder=11
    ) # 期待値の位置
    for i, coord_x in enumerate([mean_x-sd_x, mean_x+sd_x]):
        ax.axvline(
            x=coord_x, 
            color='black', linewidth=1.0, linestyle=':', 
            label=f'$\\sqrt{{V[x]}} = \\sqrt{{\\frac{{r (1-\phi)}}{{\phi^2}}}} = {sd_x:.2f}$' if i == 0 else None, 
            zorder=11
        ) # 標準偏差の位置
    ax.hlines(
        y=0.0, xmin=mean_x-sd_x, xmax=mean_x+sd_x, 
        color='black', linewidth=1.0, 
        zorder=11
    ) # 標準偏差の範囲
    ax.axvline(
        x=mode_x, 
        color='black', linewidth=1.0, linestyle='-.', 
        label=f'$mode[x] = \lfloor \\frac{{(r-1) (1-\phi)}}{{\phi}} \\rfloor = {mode_x:.2f}$', 
        zorder=11
    ) # 最頻値の位置
    ax.text(
        x=mean_x, y=0.0, 
        s='$E[x] \pm \\sqrt{V[x]}$', ha='center', va='top', 
        size=8, 
        zorder=12
    ) # 統計量のラベル
    #ax.set_xticks(ticks=x_vec) # x軸目盛
    ax.set_xlabel('$x$')
    ax.set_ylabel('probability')
    ax.set_title(param_lbl, loc='left')
    ax.legend(title='statistics', prop={'size': 8}, loc='upper left')
    ax.grid(zorder=0)
    ax.set_xlim(xmin=x_min-0.5, xmax=x_max+0.5) # (垂直線がはみ出すときの対策用)
    ax.set_ylim(ymin=y_min, ymax=y_max) # (ラベルの表示用)

# 動画を作成
anim = FuncAnimation(
    fig=fig, func=update, init_func=init, 
    frames=frame_num, interval=100
)

# 動画を書出
anim.save(
    filename='../../figure/negative_binomial/parameter/stats.mp4', 
    progress_callback=lambda i, n: print(f'\rframe: {i+1} / {n}', end='', flush=True)
)


# %%

#### パラメータとモーメントの関係 -----

# ラベルの表示用の余白を設定
y_margin = 0.05
y_min = -prob_max * y_margin
y_max = prob_max * (1.0+y_margin)

# ラベルの表示位置を設定
moment_loc_x = 0.02
moment_loc_y = 0.97

# 図を初期化
fig, ax = plt.subplots(figsize=(9, 6), dpi=100, facecolor='white')
fig.suptitle('Negative Binomial distribution', fontsize=20)

# 初期化処理を定義
def init():
    pass

# 作図処理を定義
def update(i):

    # 前フレームのグラフを初期化
    ax.cla()

    # 値を取得
    r   = r_vals[i]   # 成功回数パラメータ
    phi = phi_vals[i] # 成功確率パラメータ
    prob_vec = prob_lt[i] # 確率

    # 統計量を計算
    if phi > 0.0:
        mu    = r * (1.0-phi) / phi             # 期待値
        sigma = np.sqrt(r * (1.0-phi) / phi**2) # 標準偏差
    else: # 0除算を回避
        mu    = np.inf
        sigma = np.inf
    
    # モーメントを計算
    if phi < 1.0 and r > 0:
        skew = (2.0 - phi) / np.sqrt((1.0 - phi) * r) # 歪度
    else: # (0除算の回避用)
        skew = np.inf
    if phi > 0.0 and phi < 1.0 and r > 0:
        kurt = 6.0 / r + phi**2 / (1.0 - phi) / r # 尖度
    else: # (0除算の回避用)
        kurt = np.inf

    # ガウス分布の確率密度を計算
    norm_x_vec    = np.linspace(start=x_min, stop=x_max, num=1001)
    if sigma > 0.0: # パラメータの条件
        norm_dens_vec = norm.pdf(x=norm_x_vec, loc=mu, scale=sigma)
    else:
        norm_dens_vec = np.tile(np.nan, reps=len(norm_x_vec))
    
    # ラベル用の文字列を作成
    param_lbl   = f'$r = {r}, \\phi = {phi:.2f}, '
    param_lbl  += f'\\mu = {mu:.2f}, \\sigma = {sigma:.2f}$'
    moment_lbl  = f'skewness: {skew:.3f} \n'
    moment_lbl += f'kurtosis:    {kurt:.3f}' # (スペースによる位置調整)

    # 負の二項分布を描画
    ax.bar(
        x=x_vec, height=prob_vec, 
        color='#00A968', alpha=0.5, 
        zorder=10
    ) # 負の二項分布の確率
    ax.axvline(
        x=mu, 
        color='black', linewidth=1.0, linestyle='dashed', 
        zorder=11
    ) # 期待値の位置
    ax.hlines(
        y=0.0, xmin=mu-sigma, xmax=mu+sigma, 
        color='black', linewidth=1.0, 
        zorder=11
    ) # 標準偏差の範囲
    for label_x, label_str in zip([mu-sigma, mu, mu+sigma], ['$-\sigma$', '$\mu$', '$+\sigma$']):
        ax.text(
            x=label_x, y=0.0, 
            s=label_str, ha='center', va='top', 
            size=10, 
            zorder=12
        ) # 統計量のラベル
    ax.plot(
        x_vec, prob_vec, 
        color='#00A968', linewidth=1.0, 
        label='negative binomial', 
        zorder=13
    ) # 負の二項分布の確率
    ax.scatter(
        x=x_vec, y=prob_vec, 
        color='#00A968', s=30, 
        zorder=13
    ) # 負の二項分布の確率
    ax.plot(
        norm_x_vec, norm_dens_vec, 
        color='red', linewidth=1.0, linestyle='dashed', 
        label='gaussian', 
        zorder=14
    ) # ガウス分布の確率密度
    ax.text(
        x=moment_loc_x, y=moment_loc_y, 
        s=moment_lbl, transform=ax.transAxes, ha='left', va='top', 
        bbox=dict(facecolor='white', alpha=0.8, edgecolor='black', linewidth=0.5), 
        size = 10, 
        zorder=100
    ) # モーメントのラベル
    #ax.set_xticks(ticks=x_vec) # x軸目盛
    ax.set_xlabel('$x$')
    ax.set_ylabel('probability, density')
    ax.set_title(param_lbl, loc='left')
    ax.legend(title='distribution', prop={'size': 8}, loc='upper right')
    ax.grid(zorder=0)
    ax.set_xlim(xmin=x_min-0.5, xmax=x_max+0.5) # (垂直線がはみ出すときの対策用)
    ax.set_ylim(ymin=y_min, ymax=y_max) # (ラベルの表示用)

# 動画を作成
anim = FuncAnimation(
    fig=fig, func=update, init_func=init, 
    frames=frame_num, interval=100
)

# 動画を書出
anim.save(
    filename='../../figure/negative_binomial/parameter/moment.mp4', 
    progress_callback=lambda i, n: print(f'\rframe: {i+1} / {n}', end='', flush=True)
)


# %%

File: code/negative_binomial/parameter.R

# 負の二項分布 ---------------------------------------------------------------------

# パラメータの可視化


# ライブラリの読込 -------------------------------------------------------------

# 利用パッケージ
library(tidyverse)
library(gganimate)

# パッケージ名の省略用
library(ggplot2)


# 編集メモ ---------------------------------------------------------------------

## Φ = 0 のときにInfになり意図しない装飾になる問題に未対応


# パラメータの影響 -------------------------------------------------------------

### パラメータの設定 -----

# フレームごとのパラメータを指定
#r_vals   <- 0:100
phi_vals <- seq(from = 0, to = 1, by = 0.01)

# フレーム数を設定
#frame_num <- length(phi_vals)
frame_num <- length(r_vals)

# 固定するパラメータを作成
r_vals   <- rep(5, times = frame_num)
#phi_vals <- rep(0.5, times = frame_num)


### 変数の設定 -----

# x軸の余白を指定:(「モーメントとの関係」用)
x_margin <- 0

# x軸の範囲を設定
u <- 5
x_min <- 0 - x_margin # 余白を追加
x_max <- (r_vals * (1-phi_vals) / phi_vals) |> # 基準値を指定
  (\(.) {ifelse(test = is.infinite(.), yes = NA, no = .)})() |> # 0除算を除去
  max(na.rm = TRUE) |> 
  (\(.) {. * 0.1})() |> # 倍率を指定
  (\(.) {ceiling(. /u)*u})() |> # u単位で切り上げ
  (\(.) {. + x_margin})() # 余白を追加
x_min; x_max

# x軸の値を作成
x_vec <- seq(from = x_min, to = x_max, by = 1)


### 分布の計算 -----

# 確率を計算
anim_prob_df <- tidyr::expand_grid(
  i = 1:frame_num, # フレーム番号
  x = x_vec        # 確率変数
) |> # フレームごとに変数を複製
  dplyr::mutate(
    r    = r_vals[i],   # 成功回数
    phi  = phi_vals[i], # 成功確率
    prob = dnbinom(x = x, size = r, prob = phi) # 確率
  )


### 分布の作図 -----


#### パラメータと形状の関係 -----

# ラベル用の文字列を作成
anim_param_df <- tibble::tibble(
  i   = 1:frame_num, # フレーム番号
  r   = r_vals,      # 成功回数
  phi = phi_vals,    # 成功確率
  param_lbl = paste0(
    "list(", 
    "r == ", r, ", ", 
    "phi == ", round(phi, digits = 2), ", ", 
    "paste(psi == 1-phi, {} == ", round(1-phi, digits = 2), ")", 
    ")"
  ) # パラメータラベル
)

# 負の二項分布のアニメーションを作図
graph <- ggplot() + 
  geom_bar(
    data    = anim_prob_df, 
    mapping = aes(x = x, y = prob), 
    stat = "identity", position = "identity", 
    fill = "#00A968"
  ) + # 確率
  geom_text(
    data    = anim_param_df, 
    mapping = aes(x = -Inf, y = Inf, label = param_lbl), 
    parse = TRUE, hjust = 0, vjust = -0.5
  ) + # パラメータのラベル
  gganimate::transition_manual(frames = i) + # フレーム制御
  theme(
    plot.subtitle = element_text(size = 50) # (パラメータラベル用の空行サイズ)
  ) + # 図の体裁
  coord_cartesian(clip = "off") + # (パラメータラベル用の枠外表示設定)
  labs(
    title = "Negative Binomial Distribution", 
    subtitle = "", # (パラメータラベル用の空行)
    x = expression(x), 
    y = "probability"
  )

# 動画を作成
gganimate::animate(
  plot = graph, 
  nframes = frame_num, fps = 10, 
  width = 12, height = 8, units = "in", res = 100, 
  renderer = gganimate::av_renderer(file = "figure/negative_binomial/parameter/parameter.mp4")
)


#### パラメータと統計量の関係 -----

# ラベル用の文字列を作成
anim_param_df <- tibble::tibble(
  i   = 1:frame_num, # フレーム番号
  r   = r_vals,      # 成功回数
  phi = phi_vals,    # 成功確率
  mean_x = r * (1 - phi) / phi,         # 期待値
  sd_x   = sqrt(r * (1 - phi) / phi^2), # 標準偏差
  mode_x = ifelse(
    test = r > 1, 
    yes  = floor((r - 1) * (1 - phi) / phi), 
    no   = 1
  ), # 最頻値
  param_lbl = paste0(
    "list(", 
    "r == ", r, ", ", 
    "phi == ", round(phi, digits = 2), 
    ")"
  ), # パラメータラベル
  stat_lbl  = paste0(
    "mean: ", sprintf(fmt = '%.02f', domain = mean_x), "\n", 
    "sd:      ", sprintf(fmt = '%.02f', domain = sd_x), "\n", # (スペースによる位置調整)
    "mode: ", sprintf(fmt = '%.02f', domain = mode_x)
  ) # 統計量ラベル
)

# 標準偏差の範囲を計算
anim_sd_df <- tidyr::expand_grid(
  i    = 1:frame_num, # フレーム番号
  sign = c(-1, 1)     # 符号
) |> # パラメータごとに符号を複製
  dplyr::mutate(
    r   = r_vals[i],   # 成功回数
    phi = phi_vals[i], # 成功確率
    mean_x  = r * (1 - phi) / phi,         # 期待値
    sd_x    = sqrt(r * (1 - phi) / phi^2), # 標準偏差
    label_x = mean_x + sign*sd_x # 描画位置
  )


# ラベル用の文字列を作成
stat_lbl_vec <- c(
  mean = expression(E(x) == frac(r * (1 - phi), phi)), 
  sd   = expression(sqrt(V(x)) == sqrt(frac(r * (1 - phi), phi^2))), 
  mode = expression(mode(x) == group(lfloor, frac((r - 1) * (1 - phi), phi), rfloor))
)

# 負の二項分布のアニメーションを作図
graph <- ggplot() + 
  geom_bar(
    data    = anim_prob_df, 
    mapping = aes(x = x, y = prob), 
    stat = "identity", position = "identity", 
    fill = "#00A968"
  ) + # 確率
  geom_vline(
    data    = anim_param_df, 
    mapping = aes(xintercept = mean_x, linetype = "mean"), 
    linewidth = 1
  ) + # 期待値の位置
  geom_vline(
    data    = anim_sd_df, 
    mapping = aes(xintercept = label_x, linetype = "sd"), 
    linewidth = 1
  ) + # 標準偏差の位置
  geom_vline(
    data    = anim_param_df, 
    mapping = aes(xintercept = mode_x, linetype = "mode"), 
    linewidth = 1
  ) + # 最頻値の位置
  geom_segment(
    data    = anim_param_df, 
    mapping = aes(x = mean_x-sd_x, y = -Inf, xend = mean_x+sd_x, yend = -Inf), 
    linewidth = 1
  ) + # 標準偏差の範囲
  geom_text(
    data    = anim_sd_df,
    mapping = aes(x = label_x, y = -Inf), 
    label = "|", size = 3
  ) + # 標準偏差の指示線
  geom_label(
    data    = anim_param_df, 　
    mapping = aes(x = mean_x, y = -Inf, label = "E(x) %+-% sqrt(V(x))"), 
    parse = TRUE, hjust = 0.5, vjust = -0.2, 
    fill = "gray92", label.padding = unit(0, units = "lines"), label.size = 0, 
    size = 3
  ) + # 統計量のラベル
  geom_text(
    data    = anim_param_df, 　
    mapping = aes(x = -Inf, y = Inf, label = param_lbl), 
    parse = TRUE, hjust = 0, vjust = -0.5
  ) + # パラメータのラベル
  geom_label(
    data    = anim_param_df, 
    mapping = aes(x = -Inf, y = Inf, label = stat_lbl), 
    hjust = 0, vjust = 1, alpha = 0.5
  ) + # 統計量のラベル
  gganimate::transition_manual(frames = i) + # フレーム制御
  scale_linetype_manual(
    breaks = c("mean", "sd", "mode"), 
    values = c("dashed", "dotted", "dotdash"), 
    labels = stat_lbl_vec, 
    name   = "statistics"
  ) + # (凡例表示用)
  guides(
    linetype = guide_legend(override.aes = list(linewidth = 0.5))
  ) + # 凡例の体裁
  theme(
    plot.subtitle = element_text(size = 40) # (パラメータラベル用の空行サイズ)
  ) + # 図の体裁
  coord_cartesian(
    xlim = c(x_min, x_max), 
    clip = "off" # (パラメータラベル用の枠外表示設定)
  ) + # 描画範囲
  labs(
    title = "Negative Binomial distribution", 
    subtitle = "", # (パラメータラベル用の空行)
    x = expression(x), 
    y = "probability"
  )

# 動画を作成
gganimate::animate(
  plot = graph, 
  nframes = frame_num, fps = 10, 
  width = 12, height = 8, units = "in", res = 100, 
  renderer = gganimate::av_renderer(file = "figure/negative_binomial/parameter/stats.mp4")
)


#### パラメータとモーメントの関係 -----

# ラベル用の文字列を作成
anim_param_df <- tibble::tibble(
  i     = 1:frame_num, # フレーム番号
  r     = r_vals,      # 成功回数
  phi   = phi_vals,    # 成功確率
  mu    = r * (1 - phi) / phi,         # 期待値
  sigma = sqrt(r * (1 - phi) / phi^2), # 標準偏差
  skew  = (2 - phi) / sqrt((1 - phi) * r), # 歪度
  kurt  = 6 / r + phi^2 / (1 - phi) / r,   # 尖度
  param_lbl = paste0(
    "list(", 
    "r == ", r, ", ", 
    "phi == ", round(phi, digits = 2), ", ", 
    "mu == ", round(mu, digits = 2), ", ", 
    "sigma == ", round(sigma, digits = 2), 
    ")"
  ), # パラメータラベル
  moment_lbl = paste0(
    "skewness: ", sprintf(fmt = '%.03f', domain = skew), "\n", 
    "kurtosis:     ", sprintf(fmt = '%.03f', domain = kurt) # (スペースによる位置調整)
  ) # モーメントラベル
)

# 標準偏差の範囲を計算
anim_sd_df <- tidyr::expand_grid(
  i    = 1:frame_num, # フレーム番号
  sign = c(-1, 1)     # 符号
) |> # パラメータごとに符号を複製
  dplyr::mutate(
    r   = r_vals[i],   # 成功回数
    phi = phi_vals[i], # 成功確率
    mu      = r * (1 - phi) / phi,         # 期待値
    sigma   = sqrt(r * (1 - phi) / phi^2), # 標準偏差
    label_x = mu + sign*sigma, # 描画位置
    sd_lbl  = (sign == 1) |> 
      dplyr::if_else(
        true  = "+ sigma", 
        false = "- sigma"
      ) # 標準偏差ラベル
  )

# ガウス分布の確率密度を計算
anim_norm_df <- tidyr::expand_grid(
  i = 1:frame_num, # フレーム番号
  x = seq(from = x_min, to = x_max, length.out = 1001) # 確率変数
) |> # フレームごとに変数を複製
  dplyr::mutate(
    r     = r_vals[i],   # 成功回数
    phi   = phi_vals[i], # 成功確率
    mu    = r * (1 - phi) / phi,         # 期待値
    sigma = sqrt(r * (1 - phi) / phi^2), # 標準偏差
    dens  = dnorm(x = x, mean = mu, sd = sigma) # 確率密度
  )


# 確率密度軸の範囲を設定
dens_max <- 1

# 負の二項分布のアニメーションを作図
graph <- ggplot() + 
  geom_bar(
    data    = anim_prob_df, 
    mapping = aes(x = x, y = prob), 
    stat = "identity", position = "identity", 
    fill = "#00A968", alpha = 0.5
  ) + # 負の二項分布の確率
  geom_point(
    data    = anim_prob_df, 
    mapping = aes(x = x, y = prob), 
    color = "#00A968", size = 3
  ) + # 負の二項分布の確率
  geom_segment(
    data    = anim_param_df, 
    mapping = aes(x = mu, y = 0, xend = mu, yend = Inf), 
    linewidth = 0.8, linetype = "dashed"
  ) + # 期待値の位置
  geom_segment(
    data    = anim_param_df, 
    mapping = aes(x = mu-sigma, y = -Inf, xend = mu+sigma, yend = -Inf), 
    linewidth = 0.8
  ) + # 標準偏差の範囲
  geom_text(
    data    = anim_sd_df,
    mapping = aes(x = label_x, y = -Inf, label = "|")
  ) + # 標準偏差の指示線
  geom_text(
    data    = anim_param_df, 
    mapping = aes(x = mu, y = -Inf, label = "mu"), 
    parse = TRUE, hjust = 0.5, vjust = -0.7, 
    size = 4
  ) + # 期待値のラベル
  geom_text(
    data    = anim_sd_df, 
    mapping = aes(x = label_x, y = -Inf, label = sd_lbl), 
    parse = TRUE, hjust = 0.5, vjust = -0.7, 
    size = 4
  ) + # 標準偏差のラベル
  geom_line(
    data    = anim_prob_df, 
    mapping = aes(x = x, y = prob, color = "nb"), 
    linewidth = 1
  ) + # 負の二項分布の確率
  geom_line(
    data    = anim_norm_df, 
    mapping = aes(x = x, y = dens, color = "norm"), 
    linewidth = 1, linetype = "dashed"
  ) + # ガウス分布の確率密度
  geom_text(
    data    = anim_param_df, 
    mapping = aes(x = -Inf, y = Inf, label = param_lbl), 
    parse = TRUE, hjust = 0, vjust = -0.5
  ) + # パラメータのラベル
  geom_label(
    data    = anim_param_df, 
    mapping = aes(x = -Inf, y = Inf, label = moment_lbl), 
    hjust = 0, vjust = 1, alpha = 0.5
  ) + # モーメントのラベル
  gganimate::transition_manual(frames = i) + # フレーム制御
  scale_color_manual(
    breaks = c("nb", "norm"), 
    values = c("#00A968", "red"), 
    labels = c("Negative Binomial", "Gaussian"), 
    name   = "distribution"
  ) + # (凡例の表示用)
  guides(
    color    = guide_legend(override.aes = list(linewidth = 0.5)), 
    linetype = guide_legend(override.aes = list(linewidth = 0.5))
  ) + # 凡例の体裁
  theme(
    plot.subtitle = element_text(size = 40) # (パラメータラベル用の空行サイズ)
  ) + # 図の体裁
  coord_cartesian(
    xlim = c(x_min, x_max), 
    ylim = c(0, dens_max), 
    clip = "off" # (パラメータラベル用の枠外表示設定)
  ) + # 描画範囲
  labs(
    title = "Binomial distribution", 
    subtitle = "", # (パラメータラベル用の空行)
    x = expression(x), 
    y = "probability, density"
  )

# 動画を作成
gganimate::animate(
  plot = graph, 
  nframes = frame_num, fps = 10, 
  width = 12, height = 8, units = "in", res = 100, 
  renderer = gganimate::av_renderer(file = "figure/negative_binomial/parameter/moment.mp4")
)

warnings()


